File: w2_400/Arthur Liang/第N个数字.py
# _*_ coding: utf-8 _*_


def findNthDigit(n: int) -> int:
    digit = 1
    while n > digit * 9 * (10 ** (digit - 1)):
        n -= digit * 9 * (10 ** (digit - 1))
        digit += 1
    a = int((n - 1) / digit)
    b = int((n - 1) % digit)
    num = 10 ** (digit - 1) + a
    res = list(str(num))[b]

    return int(res[0])


if __name__ == '__main__':
    print(findNthDigit(2891))
    print(findNthDigit(111))
    print(findNthDigit(1150))
    print(findNthDigit(1600))
    print(findNthDigit(168456321154512134154))

File: w2_400/zz/第N个数字.php
class Solution {

    /**
     * @param Integer $n
     * @return Integer
     */
    function findNthDigit($n) {
        if($n<10){
            return $n;
        }
        $i= 0;
        $nmax = 0;
        while ($nmax < $n) {
            $i++;
            $nmax = $nmax + 9* pow(10,$i-1)*$i;
        }
        $num = $n-( $nmax - 9* pow(10,$i-1)*$i);
        $temp = $num / $i;
        if($num % $i == 0){
            $temp =$temp -1;
        }
        $res = $temp + pow(10,$i-1);
        return intval(strval($res)[$num % $i - 1]);
        
    }
}
